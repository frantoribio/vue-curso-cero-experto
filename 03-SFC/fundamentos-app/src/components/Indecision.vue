import Indecision from '@/components/Indecision.vue';
<template>
  <img
    src="https://via.placeholder.com/250"
    alt="bg"
  />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input
      type="text"
      placeholder="Hazme una pregunta"
      v-model.trim="question"
    />
    <p>Recuerda terminar con un signo de interrogación (?)</p>
    <div>
      <h2>{{ question }}</h2>
      <h1>Si, No, .... pensando</h1>
    </div>
  </div>
</template>

<script>
  export default {
    name: 'Indecision',
    // Mi estado
    data() {
      return {
        question: '',
      }
    },
    // Mis watchers
    watch: {
      // Observamos question
      question(newQuestion, oldQuestion) {
        // Si la pregunta cambia
        // console.log(oldQuestion, newQuestion)
        // Si la pregunta termina con un signo de interrogación
        if (newQuestion.endsWith('?')) {
          // Respuesta aleatoria
          console.log('Si tiene el signo de interrogación')
        }
      },
    },
  }
</script>

<style scoped>
  img,
  .bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
  }

  .bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
  }

  .indecision-container {
    position: relative;
    z-index: 99;
  }

  input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
  }
  input:focus {
    outline: none;
  }

  p {
    color: white;
    font-size: 20px;
    margin-top: 0px;
  }

  h1,
  h2 {
    color: white;
  }

  h2 {
    margin-top: 150px;
  }
</style>
