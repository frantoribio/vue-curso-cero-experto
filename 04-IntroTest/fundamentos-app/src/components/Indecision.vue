import Indecision from '@/components/Indecision.vue';
<template>
  <!-- Hacemos binding con image -->
  <img
    v-if="image"
    :src="image"
    alt="bg"
  />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input
      type="text"
      placeholder="Hazme una pregunta"
      v-model.trim="question"
    />
    <p>Recuerda terminar con un signo de interrogación (?)</p>
    <div>
      <h2 v-if="isValidQuestion">{{ question }}</h2>
      <h1>{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
  export default {
    name: 'Indecision',
    // eventos que emito
    emits: ['question-response'],

    // Mi estado
    data() {
      return {
        question: '',
        answer: null,
        image: '',
        isValidQuestion: false,
      }
    },

    // Mis métodos
    methods: {
      async getAnswer() {
        this.answer = 'Pensando...'
        // Obtenemos una respuesta y desestructuramos
        const response = await fetch('https://yesno.wtf/api')
        const { answer, image } = await response.json()
        this.answer = answer === 'yes' ? 'Sí' : 'No'
        this.image = image
        // Mandamos el evento y su parámetro, si lo hay
        this.$emit('question-response', this.answer)
      },
    },

    // Mis watchers
    watch: {
      // Observamos question
      question(newQuestion) {
        // Si la pregunta cambia
        // console.log(oldQuestion, newQuestion)
        // Si la pregunta termina con un signo de interrogación y tiene más que eso :)
        this.isValidQuestion = false
        if (newQuestion.endsWith('?') && newQuestion.length > 1) {
          this.isValidQuestion = true
          this.getAnswer()
        }
      },
    },
  }
</script>

<style scoped>
  img,
  .bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
  }

  .bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
  }

  .indecision-container {
    position: relative;
    z-index: 99;
  }

  input {
    width: 250px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
    margin: 1rem;
  }
  input:focus {
    outline: none;
  }

  p {
    color: white;
    font-size: 20px;
    margin-top: 0px;
  }

  h1,
  h2 {
    color: white;
  }

  h2 {
    margin-top: 150px;
  }
</style>
