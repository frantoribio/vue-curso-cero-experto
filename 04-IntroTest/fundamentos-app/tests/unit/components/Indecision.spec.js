import { describe, test, expect, beforeEach } from 'vitest'

import { shallowMount } from '@vue/test-utils'
import Indecision from '@/components/Indecision.vue'

// Suite para describir
describe('Component -> Indecision', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(Indecision)
  })

  // cada una de las pruebas o test
  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot() // snapshot es una imagen de la vista y podemos usar .element tambien o html()
  })

  test('Renders properly', () => {
    // Montamos el componente y le pasamos las propiedades
    wrapper = shallowMount(Indecision, { props: { titulo: 'Test' } })
    expect(wrapper.props().titulo).toBe('Test')
    expect(wrapper.text()).toContain('Test') // Comprobamos que el texto contiene el titulo
  })

  test('Escribir en el input no debe disparar nada', async () => {})

  test('Al escribir ? debe disparar el el fetch request', async () => {})

  test('Pruebas getAnswer', async () => {})

  test('Pruebas getAnswer - Fallo API', async () => {})

  test('Emite el evento question-response', async () => {})
})
