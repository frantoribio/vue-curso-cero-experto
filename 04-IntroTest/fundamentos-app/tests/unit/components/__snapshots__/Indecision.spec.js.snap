// Vitest Snapshot v1

exports[`Component -> Indecision > Debe hacer match con el snapshot 1`] = `
"<!-- Hacemos binding con image -->
<!--v-if-->
<div class=\\"bg-dark\\" data-v-4a8fca50=\\"\\"></div>
<div class=\\"indecision-container\\" data-v-4a8fca50=\\"\\">
  <h1 data-v-4a8fca50=\\"\\">Indecision</h1><input type=\\"text\\" placeholder=\\"Hazme una pregunta\\" data-v-4a8fca50=\\"\\">
  <p data-v-4a8fca50=\\"\\">Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <div data-v-4a8fca50=\\"\\">
    <!--v-if-->
    <h1 data-v-4a8fca50=\\"\\"></h1>
  </div>
</div>"
`;
